[docs] add StepLauncher -> Pipes migration guide (#25422)
dagster-io/dagster: docs/content/guides/migrations/from-step-launchers-to-pipes.mdx

if __name__ == "__main__":
    main()
```

Finally, provide the required resources to `Definitions`:

```python file=/guides/migrations/from_step_launchers_to_pipes/new_code.py startafter=after_downstream_marker
from dagster_aws.pipes import PipesS3MessageReader
from dagster_aws.s3 import S3PickleIOManager, S3Resource

definitions = dg.Definitions(
    assets=[upstream, downstream],
    resources={
        "pipes_emr_client": PipesEMRClient(
            client=boto3.client("emr"),
            message_reader=PipesS3MessageReader(
                client=boto3.client("s3"),
                bucket="your-bucket",
            ),
        ),
        "s3_io_manager": S3PickleIOManager(
            s3_resource=S3Resource(),
            s3_bucket="your-bucket",
            s3_prefix="your-prefix",
        ),
    },
)
```

# Conclusion

In this guide, we have demonstrated how to migrate from using step launchers to using Dagster Pipes. We have shown how to launch PySpark jobs on AWS EMR using `PipesEMRClient` and how to pass small pieces of data between assets using Dagster's metadata and Pipes extras.

# Supplementary

- [Dagster Pipes](/concepts/dagster-pipes)

> - [Dagster + Spark](integrations/spark) - an up to date list of Pipes Clients for various Spark providers can be found here

Reviewer: ```suggestion
- [Dagster + Spark](/integrations/spark) - an up to date list of Pipes Clients for various Spark providers can be found here
```

This is causing buildkite to fail as relative links are not allowed.

---
title: "Migrating From Step Launchers to Dagster Pipes | Dagster Docs"
description: "Learn how to migrate from step launchers to Dagster Pipes in Dagster."
---

> # Migrating From Step Launchers to Dagster Pipes

Reviewer: This guide is specifically demonstrating how to migrate Spark workloads.

Might there be a more general example for a migration guide, and then we can use this as a separate migrating Spark workloads guide?

Author: We don't have any Step Launchers other than Spark so that's basically a synonym

> In this guide, we’ll show you how to migrate from using step launchers to using [Dagster Pipes](/concepts/dagster-pipes) in Dagster.

Reviewer: ```suggestion
In this guide, we’ll show you how to migrate from using step launchers to using [Dagster Pipes](/concepts/dagster-pipes).
```

> The reason for this migration and a detailed comparison between the two alternatives can be found in [this GitHub Discussion](https://github.com/dagster-io/dagster/...).

Reviewer: Is this available?

Author: Not yet. I guess I'll add the exact URL once it's posted. 

Or should we create the GitHub Discussion before merging this guide? @yuhan

Author: Decided to remove this for now

---

## Considerations

When deciding to migrate from step launchers to Dagster Pipes, consider the following:

- **Step launchers** are superceded by Dagster Pipes. While they are still available (and there are no plans for their removal), they are no longer the recommended method for launching external code from Dagster ops and assets. They won't be receiving new features or be under active development.
- **Dagster Pipes** is a more lightweight and flexible framework, but it does come with a few drawbacks:
  1. Spark runtime and the code executed will no longer be managed by Dagster for you.
  2. Dagster Pipes are not compatible with Resources and IO Managers. If you are heavily relying on these features, you might want to keep using step launchers.

## Steps

To migrate from step launchers to Dagster Pipes, you will have to perform the following steps.

> ### **1. Implement new CI/CD pipelines to prepare your Spark runtime environment**

Reviewer: Is this step necessary? Personally, I found it a bit overwhelming.

Alternatively, could we just state that the user must install `dagster-pipes` in their execution environment?

Author: Previously the Step Launcher was doing all this setup including package preparation for the user. 

I know from personal experience how hard it is to get AWS EMR working so IMO these concrete steps are nice to have, but we can also just link existing EMR Pipes docs.

```python file=/guides/migrations/from_step_launchers_to_pipes/old_code.py
from typing import Any

from dagster_aws.emr import emr_pyspark_step_launcher
from my_lib import MyPysparkIOManager, calculate_metric, get_data_frame
from pyspark.sql import DataFrame

import dagster as dg

# the upstream asset will serve as an example of writing a Spark DataFrame

@dg.asset(io_manager_key="pyspark_io_manager")
def upstream(pyspark_step_launcher: dg.ResourceParam[Any]) -> DataFrame:
    return get_data_frame()

# the downstream asset will serve as an example of reading a Spark DataFrame
# and logging metadata to Dagster

@dg.asset(io_manager_key="pyspark_io_manager")
def downstream(
    context: dg.AssetExecutionContext,
    upstream: DataFrame,
    pyspark_step_launcher: dg.ResourceParam[Any],
) -> None:
    my_metric = calculate_metric(upstream)
    context.add_output_metadata({"my_metric": my_metric})
    return

definitions = dg.Definitions(
    assets=[upstream, downstream],
    resources={
        "pyspark_step_launcher": emr_pyspark_step_launcher,
        "pyspark_io_manager": MyPysparkIOManager(),
    },
)
```

> The corresponding Pipes code will be split in 2 pieces: the Dagster part and the PySpark job part.

Reviewer: ```suggestion
The corresponding Pipes code will instead have two components: the Dagster asset definition, and the external PySpark job.
```

Let's start with the PySpark job. The upstream asset will invoke the following script:

```python file=/guides/migrations/from_step_launchers_to_pipes/upstream_asset_script.py
import os

import boto3
from dagster_pipes import PipesS3MessageWriter, open_dagster_pipes

> from my_lib import get_data_frame

Reviewer: Can we simplify this by instantiating a SparkSession, and not reference a fictional `my_lib`?

Author: I think the point of this is to clearly show how the business logic is being moved to the external script, I would rather keep it this way

def main():
    with open_dagster_pipes(
        message_writer=PipesS3MessageWriter(client=boto3.client("s3")),
    ) as pipes:
        df = get_data_frame()

        path = "s3://" + os.path.join(*pipes.asset_identifier) + ".parquet"

        # this was probably previously logged by the IOManager
        pipes.add_output_metadata({"path": path})

        df.write.parquet(path)

if __name__ == "__main__":
    main()
```

Now, we have to run this script from Dagster. First, let's factor the boilerplate EMR config into a reusable function:

```python file=/guides/migrations/from_step_launchers_to_pipes/utils.py startafter=start_emr_config_marker endbefore=end_emr_config_marker

> def make_emr_params(script_path: str) -> dict:

Reviewer: It's nice to show a real-world example of using EMR, but in the earlier example we're not using EMR with the Step Launcher.

Would this example be simplified if we used local execution mode for Spark?

Author: We don't really have Pipes for local execution. 

> in the earlier example we're not using EMR with the Step Launcher.

We are, the EMR Step Launcher is passed to Definitions
